### DISCUSSION TITLE
Custom Service Routes beyond CRUD (without using X-Service-Method)

### DISCUSSION THREAD

**Discussion by dustinc** [feathersjs/feathers#3423] [Q&A] [answered]
I am loving Dove so far, but am running into a kind of major roadblock while attempting something super simple.

The project is an existing REST API that I am converting from Express (JS) to Feathers (TS) w/ a switch to Koa. I prefer Koa anyways. I have a resource `Gig` that does most of the heavy lifting. There are many endpoints beyond basic CRUD operations e.g. GET `/gigs/requests/pending` (gets all pending requests for all gigs for admin users to see),  GET `/gigs/:id/requests` (gets all requests for specific gig), POST `/gigs/:id/confirm` (lets user confirm their gig) and so on. In Express and Koa you simple stack your routes in order to prevent "not an id" error, but you're free to add whatever routes you wish.

I've read about Custom Service Method and was only able to find a solution that, for me, would equate to POST `/gigs` with header `X-Service-Method=xxxx`. This may be fine for greenfield projects (I've never used this approach before in a REST API), but I would like to make this transition with minimal to zero impact on the mobile app that is consuming this API.

Why is this so difficult to achieve in Feathers Dove? I really really really want to continue using Dove, but I cannot seem to find any resources regarding a solution, and I would prefer not to hack in something. What is the "feathers" way to solve this fairly basic problem?

Hopefully I am just missing something really simple that someone can point out to me. (feel free to point and laugh afterwards 😄, I'll welcome it!)

Please help!! Thank you.

**Comment by daffl** [comment, marked as answer]
You can create any route structure you need with placeholders and custom services. In this case e.g.

```ts
class GigConfirmService {
  constructor(app) {
    this.app = app
  }

  async create (data, params) {
    const gigId = params.route.id
    return this.app.service('gigs').patch(gigId, { status: 'confirmed' }) 
  }
}

class GigRequestService {
  constructor(app) {
    this.app = app
  }

  async find (params) {
    const gigId = params.route.id
    return this.app.service('gigs').find({
      query: { gigId }
    })
  }
}

app.use('gigs/:id/confirm', new GigConfirmService(app))
app.use('gigs/:id/requests', new GigRequestService(app))
```

How service methods map to HTTP methods is documented in the [REST API client docs](https://feathersjs.com/api/client/rest.html#http-api).

**Reply by dustinc** [reply]
This is unfortunately a deal breaker. I did not realize that services were hand-tied in this manner. Such a disappointment; I was having so much fun with Feathers until this. Thanks anyways.

**Reply by daffl** [reply]
What I was getting it is that a service is just what any other middleware would do. You can also re-register an existing service for nested routes and map the route parameters in a hook. This is [explained in more detail in the FAQ](https://feathersjs.com/help/faq.html#how-do-i-do-nested-or-custom-routes).

**Comment by dustinc** [comment]
@daffl Thank you so much for your response; I will try this out right now. It does feel strange creating a new service for specific endpoints on the same resource. I was really hoping to map a custom route to a custom service method all within the same resource-specific service.

Thank you!

**Reply by Mairu** [reply]
As a side note, with feathers-swagger it is possible to setup custom REST routes for custom methods: https://feathersjs-ecosystem.github.io/feathers-swagger/#/custom-methods

**Reply by dustinc** [reply]
This effectively solves my problem. Thank you for sharing!
